acnportal/acnsim/models/battery.py

```diff
@@ -188,3 +284,109 @@ def _from_dict(cls, attribute_dict, context_dict, loaded_dict=None):
+                      max_rate=32, transition_soc=0.8):
+        """ Given a requested energy, stay duration, battery capacity
+        (kWh), charging voltage, period, max charging rate, and the
+        state of charge (SoC) at which non-ideal behavior begins, finds
+        the maximum initial capacity the battery may have such that
+        requested_energy is delivered in stay_dur periods, assuming
+        linear 2 stage battery behavior.
+
+        """
+        delta_soc = requested_energy / cap
+        # Maximum rate of change of SoC in per period.
+        max_dsoc = max_rate * voltage / 1000 / cap / (60 / period)
+
+        # First, assume init_soc is after the transition_soc. In this
+        # case, the max init_soc has a closed form solution.
+        init_soc = (
+            1
+            + delta_soc
+            / (np.exp(max_dsoc * stay_dur / (transition_soc - 1)) - 1)
+        )
+        if init_soc >= transition_soc:
+            return init_soc
+
+        # If that didn't work, search over all possible init_soc for the
+        # largest init_soc that still allows for delta_soc to be
+        # delivered in stay_dur periods.
+        def delta_soc_from_init_soc(init_soc):
+            if stay_dur <= (transition_soc - init_soc) / max_dsoc:
+                return max_dsoc * stay_dur
+            return (
+                1
+                + np.exp(
+                    (max_dsoc * stay_dur + init_soc - transition_soc)
+                    / (transition_soc - 1)
+                )
+                * (transition_soc - 1)
+                - init_soc
+            )
+
+        # Before that, we make sure that starting at init_soc of 0, it's
+        # possible to deliver the requested energy with this battery.
+        if delta_soc_from_init_soc(0) < delta_soc:
+            return -1
+
+        # Since max energy delivered is decreasing in init_soc, we
+        # use a binary search for a decreasing function. The default
+        # tolerance is set to 1e-9 to satisfy the tests'
+        # assertAlmostEqual default tolerance.
+        def binsearch(f, lb, ub, targ, tol=1e-9):
+            mid = (lb + ub) / 2
+            val = f(mid)
+            if abs(val - targ) < tol:
+                return mid
+            elif val - targ > 0:
+                return binsearch(f, mid, ub, targ, tol=tol)
+            else:
+                return binsearch(f, lb, mid, targ, tol=tol)
+
+        # Below an initial capacity of
+        # (transition_soc - max_dsoc * stay_dur),
+        # the same amount of charge is delivered, as charging stays
+        # entirely within the ideal region. So, we can start the binary
+        # search here.
+        init_soc = binsearch(
+            delta_soc_from_init_soc,
+            transition_soc - max_dsoc * stay_dur,
+            1, delta_soc
+        )
+        return init_soc * cap
+
+    # Potential capacities taken from TODO: find source.
+    potential_caps = np.array([8, 24, 40, 60, 85, 100])
```

Comment:
I had meant to ask you about this. Do you have a source that lists these capacities as the most common ones?

Reply:
I'll add this to issue #60